ocaml/xapi/xapi_xenops.ml

```diff
@@ -2029,48 +2025,107 @@ let on_xapi_restart ~__context =
+    "The following VMs are running in xenopsd that xapi does not know about"
+    xenopsd_vms_not_in_xapi;
+
+  maybe_log_em
+    "The following VMs are running in xenopsd but xapi thinks are running elsewhere."
+    xapi_thinks_are_elsewhere; (* This is bad! *)
+
+  maybe_log_em
+    "The following VMs are running in xenopsd but xapi thinks are running nowhere."
+    xapi_thinks_are_nowhere; (* This is pretty bad! *)
+
+  if List.length xapi_vms_not_in_xenopsd > 0 then begin
+    debug "The following VMs are not running in xenopsd, but xapi thought they were";
+    List.iter (fun (id,_) -> debug "%s" id) xapi_vms_not_in_xenopsd
+  end;
+
+  (* Destroy any VMs running that aren't in Xapi's database, or that xapi
+     thinks are running on another host *)
   List.iter (fun ((id, state), queue_name) ->
       let module Client = (val make_client queue_name : XENOPS) in
-      info "VM %s is running here but isn't in the database: terminating" id;
+      info "VM %s is running here but isn't supposed to be: terminating" id;
       if state.Vm.power_state <> Halted then
         Client.VM.shutdown dbg id None |> wait_for_task queue_name dbg |> ignore;
       Client.VM.remove dbg id
-    ) xenopsd_vms_not_in_xapi;
+    ) (xenopsd_vms_not_in_xapi @ xapi_thinks_are_elsewhere);
 
-  (* Sync VM state in Xapi for VMs running by local Xenopsds *)
+  (* Sync resident_on state in Xapi for VMs running by local Xenopsds that
+     xapi didn't think were anywhere *)
   List.iter (fun ((id, state), queue_name) ->
       let vm = vm_of_id ~__context id in
-      let xapi_power_state =
-        xenapi_of_xenops_power_state (Some state.Vm.power_state) in
-      Xapi_vm_lifecycle.force_state_reset ~__context ~self:vm ~value:xapi_power_state;
-      match xapi_power_state with
-      | `Running | `Paused ->
-        Db.VM.set_resident_on ~__context ~self:vm ~value:localhost;
-        add_caches id;
-        refresh_vm ~__context ~self:vm;
-      | `Suspended | `Halted ->
-        let module Client = (val make_client queue_name : XENOPS) in
-        Client.VM.remove dbg id;
-        if List.exists (fun (id', _) -> id' = id) resident_vms_in_db
-        then Db.VM.set_resident_on ~__context ~self:vm ~value:Ref.null;
-    ) xenopsd_vms_in_xapi;
+      info "Setting resident_on for VM %s to be this host" id;
+      Db.VM.set_resident_on ~__context ~self:vm ~value:localhost)
+      xapi_thinks_are_nowhere;
 
   (* Sync VM state in Xapi for VMs not running on this host *)
   List.iter (fun (id, vm) ->
+      info "VM %s was running in the DB but isn't any more. Resetting in DB" id;
       Xapi_vm_lifecycle.force_state_reset ~__context ~self:vm ~value:`Halted;
       Db.VM.set_resident_on ~__context ~self:vm ~value:Ref.null;
-    ) xapi_vms_not_in_xenopsd
+  ) xapi_vms_not_in_xenopsd
+
+let resync_all_vms ~__context =
+  (* This should now be correct *)
+  let localhost = Helpers.get_localhost ~__context in
+  let resident_vms_in_db =
+    List.filter (fun self ->
```

Comment:
Localhost is used elsewhere in the function so I've left that bit, but I have restructured a bit as you have suggested.